Add YR Civil Structures to all theaters YR and RA2

CCCP84 opened the pull request:
I did a lot of work identifying all the buildings that weren't displayed in the various theaters in YR and added them to all the theaters.
 I also added the North Korean flag and Barrier, which weren't used before. There is a small problem with the barrier: the barrier graphics is displayed over the graphics of the vehicles driving over it. Maybe someone knows something. I tried flat=Yes but it didn't work.

I couldn't find graphics for several buildings.
 
Tomorrow i hope I'll work on the RA2 mode.

It is very convenient to use these maps for testing:
[BUILDINGS TEST.zip](https://github.com/user-attachments/files/23615188/BUILDINGS.TEST.zip)


https://github.com/user-attachments/assets/c1ec0257-4ca9-47d3-abae-0eccb203d46d


https://github.com/user-attachments/assets/e8751c47-809c-44c6-8045-0be14c43e3f8

<img width="218" height="227" alt="image" src="https://github.com/user-attachments/assets/411fcb19-9032-40ea-9353-2f6cb9372e17" />

<img width="1920" height="1200" alt="image" src="https://github.com/user-attachments/assets/ebf48816-d3a5-43c2-9aa2-8e830f4e0205" />

<img width="1920" height="1200" alt="image" src="https://github.com/user-attachments/assets/3d545567-714c-47ad-b509-77611405ab53" />

CCCP84 committed:
Adding all buildings to all theaters YR

CnCRAZER commented:
Nice job including the yuri command center. No telling how many times I've tried to use it in certain theaters but turned out invisible. Looking forward to this one 👍🏻

CCCP84 committed:
Small HOLLYWOOD sign

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update artmd.ini

Belonit commented:
For now I'll mark this as draft, since the rules and art are not updated for the RA2 mod

CCCP84 committed:
YR buildings to RA2

CCCP84 committed:
Update artmd.ini

CCCP84 committed:
Update artmd.ini

CCCP84 commented:
I added the same set of buildings to Red Alert except Yuri's buildings. This can be done in another PR if we want it

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update rulesmd.ini

CCCP84 commented:
The pyramids will finally appear in Tour of Egypt in Red Alert..

Damn, adding all these buildings took a really long time

CnCRAZER commented:
> The pyramids will finally appear in Tour of Egypt in Red Alert..
> 
> 
> 
> Damn, adding all these buildings took a really long time

It's a lot of work for sure. 😆

CCCP84 committed:
Add Watchtower!

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update rulesmd.ini

CCCP84 commented:
Found, adjusted and added a watchtower 👍👍👍 
I think I'll use this somewhere in my maps

<img width="200" height="211" alt="image" src="https://github.com/user-attachments/assets/374bf5d9-b542-42ad-bfde-8475836213b4" />

CCCP84 committed:
Add GASPOT

MahBoiDeveloper commented on game-assets/cncnet.pack/rulesmd.ini:
```diff
@@ -22594,7 +22594,7 @@ DieSound=BuildingMetalDamaged
 ;NK Flag Left; CCCP84 Nov 2025: it was in unused files. added to ini files and shp to the mix
 [CANKFGL]
-UIName=Name:NKOREAFLAG
+UIName=NOSTR:N.Korean Flag
```
No, `NOSTR` strings are unacceptable for this config. If you want to add new strings, add them to the .csf file instead.

MahBoiDeveloper commented on game-assets/cncnet.pack/rulesmd.ini:
```diff
@@ -21416,6 +21468,46 @@ CanBeOccupied=no
 
+;Watchtower ; CCCP84 Nov 2025: It was in unused files. Added to ini files and shp to the mix. Added Muzzle Flashes. Wow, cool thing )
+[CAGARD03]
+UIName=NOSTR:Watchtower
```
Replace `NOSTR` with the regular string.

MahBoiDeveloper commented on game-assets/ra2mode.pack/rulesmd.ini:
```diff
@@ -13211,197 +15348,225 @@ RadarInvisible=yes
+
+;Watchtower ; CCCP84 Nov 2025: It was in unused files. Added to ini files and shp to the mix. Added Muzzle Flashes. Wow, cool thing )
+[CAGARD03]
+UIName=NOSTR:Watchtower
```
Replace `NOSTR` with the regular string.

MahBoiDeveloper commented on game-assets/ra2mode.pack/rulesmd.ini:
```diff
@@ -15030,1078 +17198,1206 @@ ClickRepairable=no
-Name=RA2 Harvard's Office
+;NK Flag Left; CCCP84 Nov 2025: it was in unused files. added to ini files and shp to the mix
+[CANKFGL]
+UIName=NOSTR:N.Korean Flag
```
Replace `NOSTR` with the regular string.

CCCP84 commented:
@MahBoiDeveloper where is file with UInames? to give name for new object?
However, it is not necessary to do this.

CCCP84 commented:
I think it could be a merged, and preferably faster so there is time for testing.
@CnCRAZER @Metadorius

CCCP84 commented:
Light Tower

<img width="300" height="306" alt="image" src="https://github.com/user-attachments/assets/e6b082d2-4120-438e-bab8-a88ca4f09f29" />

MahBoiDeveloper commented:
> where is file with UInames? to give name for new object? However, it is not necessary to do this.
> I think it could be a merged, and preferably faster so there is time for testing.

The `NOSTR` strings in `UIName` cannot be replaced by strings from `stringtableXX.csf` or `ra2md.csf`. You should not use them in `rulesmd.ini` because you block `UIName` to be translated in future.

CCCP84 commented:
@MahBoiDeveloper  So you're suggesting adding a CSF files for the corresponding mixes? Do we really need it, when now we can add names to rules? Existing names could be read from the CSF, and new names can be added to rules.
What are the possible disadvantages of this scheme? Now it's simplified.

MahBoiDeveloper commented:
> So you're suggesting adding a CSF files for the corresponding mixes?

Such CSF file already exists in CnCNet's mix files: https://github.com/CCCP84/cncnet-yr-client-package/blob/Buildings_all_theaters_YR/game-assets/cncnet.pack/stringtable00.csf

CCCP84 commented:
but there is no for RA2 mode, when it comes out, we'll transfer the name there. and in general we still need to think about it.

CnCRAZER commented:
Since you're adding multiple building types to the list let's change this a little.
Currently:
```407=CALA03A; CCCP84 Nov 2025: Complete small "HOLLYWOOD" sign. Added to all theaters
408=CANKFGL; CCCP84 Nov 2025: North Korean Flag. It was in unused files. Added to ini files and shp to the mix
409=CAGARD03; CCCP84 Nov 2025: Watchtower. It was in unused files. Added to ini files and shp to the mix. Added Muzzle Flashes. Wow, cool thing )
410=GAGATE_B; CCCP84 Nov 2025: It was in unused files. Added to ini files, fixed shp and added it to the mix. The building's graphics is displayed over units passing over it
413=GACTWR
414=GASPOT; CCCP84 Nov 2025: It was in unused files. Added to rules.ini, fixed shp and added it to the mix.
```

I propose something bit more easier to read:
```
; CCCP84 - Restored buildings
407=CALA03A; Hollywood Sign
408=CANKFGL; North Korean Flag
409=CAGARD03; Watch Tower
410=GAGATE_B; Leftover gate
413=GACTWR
414=GASPOT; Light Tower

CnCRAZER commented:
> but there is no for RA2 mode, when it comes out, we'll transfer the name there. and in general we still need to think about it.

If any of these buildings don't have a valid csf string already in the mix files, perhaps we could find one that already exists that is similar to re-use? For translation purposes of course

CCCP84 committed:
Add new buildings names

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update rulesmd.ini

CnCRAZER reviewed (dismissed):
So far this looks good. Approving while on mobile but will take a second look when at computer 👍🏻

MahBoiDeveloper reviewed (commented):
Need to check in-game tooltips for new structures. 

@CCCP84, would you kindly provide your test map, so we can sure are the new additions is good to merge.

@Belonit `stringtable00.csf` changes affects the RA2 game mod, or only YR?

CCCP84 commented:
> @CCCP84, would you kindly provide your test map, so we can sure are the new additions is good to merge.

In first post

CCCP84 commented:
> @Belonit `stringtable00.csf` changes affects the RA2 game mod, or only YR?

Both

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update rulesmd.ini

CCCP84 committed:
Update rulesmd.ini

SadPencil reviewed (changes requested):
Please do not modify `stringtable00.csf` file as it is intentionally reserved for Ares and Phobos strings. You can create a `stringtable01.csf` file instead. Also, please also explictly list what's inside in the new CSF file (you can do so by using [Ra2CsfToolsGUI](https://github.com/SadPencil/Ra2CsfToolsGUI/releases/))

SadPencil commented:
> @MahBoiDeveloper where is file with UInames? to give name for new object? However, it is not necessary to do this.

We don't have such a file. However, `stringtable01.csf` should be fine for this purpose

CCCP84 commented:
> Пожалуйста, не изменяйте `stringtable00.csf`файл, так как он намеренно зарезервирован для строк Ареса и Фобоса. Вы можете создать новый `stringtable01.csf`файл. Также, пожалуйста, укажите содержимое нового CSF-файла (это можно сделать с помощью [Ra2CsfToolsGUI] ).

Why shouldn't you use this file? What are the possible consequences?

SadPencil commented:
> Why shouldn't you use this file? What are the possible consequences?

Same as why there are `stringtable00` -- `stringtable99` csf files instead of only one, and why programming codes are written in multiple files and folders instead of only one file.

stringtable00 contains Ares and Phobos strings, which is different from in-game assets. The English version is a copy from `Ares.mix` but translators also need to manually add and follow up Phobos strings in the translation of this file. If you merge this file with in-game strings, this means translators need to track on two upstream sources -- Phobos and yr package, introducing chaos.

CCCP84 committed:
Separate csf

CCCP84 commented:
<img width="506" height="458" alt="image" src="https://github.com/user-attachments/assets/8aca6925-d49f-409b-929d-5a7d7c96abcc" />

SadPencil commented:
Hit: you can use [Ra2CsfToolsGUI](https://github.com/SadPencil/Ra2CsfToolsGUI/releases/) to paste the .ini file to represent the whole content of CSF file, since a screenshot only shows a single CSF value. I'll do it for you once.

```
[SadPencil.Ra2CsfFile.Ini]
IniVersion=2
CsfVersion=3
CsfLang=0

[Name:CANKFGL]
Value=N.Korean Flag

[Name:CAGARD03]
Value=Watchtower
```

<img width="1700" height="1100" alt="" src="https://github.com/user-attachments/assets/c1667927-ad43-4cf2-b422-30566f4511e5" />

SadPencil commented:
`N.Korean Flag`? is this a common way to express this country? Perhaps `North Korean Flag` or `DPRK Flag`

Besides I doubt whether North Korean exist or not at all in the story of Red Alert 2, since it's based on a fiction story that differs after 1924. This might explains why this flag is not used before -- conflict with the RA2 story.

SadPencil commented:
Besides, I don't have other concerns. So if I did not submit an updated review comment, feel free to dismiss my 'requesting changes' review upon finishing the discussion of this flag question. I can accept it either updating it with a commonly used name, or remove the usage of this flag.

CCCP84 commented:
I don't think relevance to story of Red Alert 2 is a factor in deciding whether to include the building. This flag can be used by players to create their own maps.
 I also don't think we need to be absolutely precise in the country's name. I think it's enough that most players will understand what we're talking about. I just thought the name "North Korean flag" was too long for a pop-up title, and a shortened abbreviation would be confusing. I like the current version. But we can replace it with the "North Korean Flag"

SadPencil commented:
We have


[Name:CAMSC01]
Value=Mr. Yummy's Hot Dogs!

so length should not be a problem

DeathFishAtEase reviewed (changes requested):
There are the following correction/improvement plans:
- [x] 01. `cggard03.shp`: There are already six theater-specific Shape files, so `cggard03.shp` is not needed. I believe this is an error:
    - Otherwise, we would have to create versions with the second letter as `g` for many files like `c?outp.shp`.
    - And doing so would also be inconsistent with your handling of `c(d/l)seat01.shp`.
- [x] 02. `cgla03a.shp`: This appears to be a rename of `cala03.shp`. Since we now have the snow version of the Hwood Sign as a separate file, perhaps we can similarly replace the original `cala03.shp` with a version consistent with other theaters? Of course, this would require some changes at the map level. I did a quick count: currently, `cala03.shp` is used by two regular maps among our `[Map] -> Theater=SNOW` maps, one of which also has a blitz version:
    - `2_sen_river.map`
    - `2_svalbard.map` & `blitz_svalbard_2v2.map`
- [x] 03. `cgnkfgl.shp`: There are already six theater-specific Shape files, so `cgnkfgl.shp` is not needed.
- [x] 04. `cgnkfgl_a.shp`: Same as above.
- [x] 05. `ctsanf04.shp`: The same smearing issue also occurs with `cusanf04.shp` and `clsanf04.shp`. It is recommended to handle them together.
- [x] 06. `cgsanf05.shp`: I want to confirm whether `casanf05.shp` does not need processing, even though this BuildingType is only used in `all/sov01umd.map` and both are `[Map] -> Theater=NEWURBAN` rather than `SNOW`.
- [x] 07. `ctsanf09.shp`: The same *smearing* issue also occurs with `clsanf09.shp`. It is recommended to handle them together.
- [x] 08. `ctsanf10.shp`: The same *smearing* issue also occurs with `clsanf10.shp`. It is recommended to handle them together.
- [x] 09. `ctsanf11.shp`: The same *smearing* issue also occurs with `clsanf11.shp`. It is recommended to handle them together.
- [x] 10. `ctsanf12.shp`: The same *pattern* issue also occurs with `clsanf12.shp`. It is recommended to handle them together.
- [x] 11. `ctsanf13.shp`: The same *pattern* issue also occurs with `clsanf13.shp`. It is recommended to handle them together.
- [x] 12. `ctsanf14.shp`: The same *pattern* issue also occurs with `clsanf14.shp`. It is recommended to handle them together.
- [x] 13. `cgsanf15.shp`: The snow version of this asset, `casanf15.shp`, has an error in the third frame and requires additional processing. This BuildingType is currently used in the following maps:
    - With `[Map] -> Theater=URBAN`:
      - `blitz_jurbarkcity.map`
      - `hillbtwn.map`
      - `8_andalusia.map`
    - With `[Map] -> Theater=NEWURBAN`:
      - `c4w01md.map`
      - `manhatta.map`
      - `8_invasion.map`
      - `8_octo_bay.map`
      - `8_pipeline_oil_in_center.map`
    - Based on usage, the snow version of this BuildingType art asset can be safely changed.
- [x] 14. `cgsanf16.shp`: The snow version of this asset, `casanf16.shp`, actually uses the image of `casanf15.shp` for the main body and the first half of the shadow (frames 1-4, 9-12 in 1-based indexing). Additional processing is required.
    - Currently, among the maps we use, only `all/sov01umd.map` and `8_octo_bay.map` (3 maps with `[Map] -> Theater=NEWURBAN`), `2_circuit_board.map` and `sonichighways.map` (2 maps with `[Map] -> Theater=URBAN`), and `8_malaboo_bay.map` (1 map with `[Map] -> Theater=TEMPERATE`) contain this BuildingType. Therefore, the snow version can be safely changed.
- [x] 15. `cgsanf17.shp`: The snow version of this asset, `casanf17.shp`, actually has a reduced image size and requires additional processing.
    - Currently, among the maps we use, only `blitz_jurbarkcity.map` and `sonichighways.map` (2 maps with `[Map] -> Theater=URBAN`) and `all/sov01umd.map` (2 maps with `[Map] -> Theater=NEWURBAN`) contain this BuildingType. Therefore, the snow version can be safely changed.
- [x] 16. `cgsanf18.shp`: The snow version of this asset, `casanf18.shp`, actually has a reduced image size and requires additional processing.
    - Currently, among the maps we use, only `all/sov01umd.map`, `all02umd.map`, and its derivatives `c1a02md.map` and `rushhr.map` (totaling 5 maps with `[Map] -> Theater=NEWURBAN`) contain this BuildingType. Therefore, the snow version can be safely changed.
- [x] 17. `cdtran03.shp` & `cltran03.shp`: I'm glad to see them corrected, but since the original game lacks `cutran03.shp`, there are still some unresolved issues. It is recommended to handle them together.
- [x] 18. `gggate_b.shp`: This is a leftover from Tiberian Sun, only it was enlarged to fit the dimensions of RA2/YR, but in fact, in TS it was used for GDI's gate. Under the same Theater, there are 4 files respectively `g?gate_a.shp` and `g?gateza.shp` and symmetric versions ending with `b.shp`. In RA2/YR, `*_a.shp`/`*za.shp` is used as the new gate, therefore I think reusing TS assets is not a correct approach. A better solution is to replace `*zb.shp` with the symmetric version of `*za.shp`.
    The style of `gggateza.shp` in RA2:<img width="176" height="127" alt="image" src="https://github.com/user-attachments/assets/1fd9c5a5-a169-4111-be53-b030fc22ec24" />
- [x] 19. `ggspot.shp`: This asset already exists in the original `ra2(md).mix` with the same style. For what purpose does it need to be added to `cncnet.mix`?
    - Additionally, it is a leftover from Tiberian Sun. Do we have a need to handle such files? If so, that seems to go a bit far.
    - As mentioned, it is inherited from Tiberian Sun, so the snow version of this asset, `gaspot.shp`, actually has a reduced image size (in fact, it retains the size from Tiberian Sun, so it is "reduced" compared to the non-snow version). Additional processing is required.

CCCP84 commented:
Everything that's written is complete nonsense. I've corrected the added animations. I don't even understand what the rest is about.
You're writing some nonsense about me adding certain animations that I didn't add.
If I added some animations that existed, it means that I fixed them, they were corrupted.

CCCP84 commented:
@DeathFishAtEase I suggest that after this PR is merged, you create your own PR and propose your changes.

CCCP84 commented:
Well, some points make sense. Regarding the North Korean flag and cggard03, there really seem to have shp files.

CCCP84 committed:
remove unneccessary shps

CCCP84 commented:
@DeathFishAtEase Can you explain what you mean in points 5-13? and 17.

CCCP84 commented:
@DeathFishAtEase  Okay, I'll probably rephrase my conclusion a little. Many of these points make sense, but the meaning of some of them is not clear for me )

CCCP84 commented:
I don't understand what should be done with the reduced-size images of buildings. In your opinion, they should be replaced? The Hollywood sign can also be considered a reduced size. Perhaps that was the intention.

CCCP84 commented:
Tell me again about number 14. Do you think it can be replaced? Is it not used anywhere?

CCCP84 commented:
Regarding the reduced Hollywood inscription and replacing it in the maps, I think someone should create another PR. I'm already tired of this.

CCCP84 commented:
> 5. `ctsanf04.shp`: The same smearing issue also occurs with `cusanf04.shp` and `clsanf04.shp`. It is recommended to handle them together.

I think I understand what you meant. Do you meant the "D" letter?

CCCP84 committed:
add bridge shps without "D" issue

CCCP84 commented:
Apparently that's what you meant. You was right and I was wrong.
I added files without "D" issue, however, for clsanf12,13,14 the problem was not confirmed in my tests

CCCP84 commented:
> 13\. `cgsanf15.shp`: The snow version of this asset, `casanf15.shp`, has an error in the third frame and requires additional processing.

i didn't notice any problems with it. What error do you mean?

CCCP84 committed:
cutran03

CCCP84 commented:
> 18\. `gggate_b.shp`: This asset already exists in the original `ra2(md).mix` with the same style. For what purpose does it need to be added to `cncnet.mix`?

> * `gggate_b.shp`: This asset already exists in the original `ra2(md).mix` with the same style. For what purpose does it need to be added to `cncnet.mix`?

i fixed and added theese shps

SadPencil commented:
> I don't think relevance to story of Red Alert 2 is a factor in deciding whether to include the building. This flag can be used by players to create their own maps. I also don't think we need to be absolutely precise in the country's name. I think it's enough that most players will understand what we're talking about. I just thought the name "North Korean flag" was too long for a pop-up title, and a shortened abbreviation would be confusing. I like the current version. But we can replace it with the "North Korean Flag"

Can you update the CSF with "North Korean Flag" then?

> To https://github.com/CCCP84/cncnet-yr-client-package.git
 ! [remote rejected]   pr/758 -> Buildings_all_theaters_YR (permission denied)
error: failed to push some refs to 'https://github.com/CCCP84/cncnet-yr-client-package.git'

DeathFishAtEase committed:
Adaptation issues of small/large Hollywood signs

Change the SNOW version of `cala03.shp` to be consistent with other theaters.

Map changes required to adapt to the migration of `[CALA03]` and `[CALA03A]`:
- Since their Foundation is consistent, they can be directly replaced.
For more information, see <https://github.com/CnCNet/cncnet-yr-client-package/pull/758#pullrequestreview-3505363359>.

DeathFishAtEase commented:
> I suggest that after this PR is merged, you create your own PR and propose your changes.

>> 13. `cgsanf15.shp`: The snow version of this asset, `casanf15.shp`, has an error in the ***third*** frame and requires additional processing.
> 
> i didn't notice any problems with it. What error do you mean?

> Tell me again about number 14. Do you think it can be replaced? Is it not used anywhere?

I have not found any map with `[Map] -> Theater=SNOW` that contains `[CASANF16]`, so theoretically it can be safely replaced.
The simplest explanation of its issue is inconsistency with versions from other theaters (It will use the same images as the snow version `[CASANF15]`, and has a higher logical height and selection box), but the deeper problem lies in:
- The sequence of `ctsandf15.shp` is A,B,A,C,A(S),B(S),A(S),C(S), where (S) represents the shadow frame.
- The incorrect sequence of `casanf15.shp` is: A,B,***B***,C,A(S),B(S),***B(S)***,C(S)
  - The third frame of `[CASANF15]` is inconsistent with the versions of other theaters.
- The correct sequence of `ctsanf16.shp` is: D,E,D,F,D(S),E(S),D(S),F(S)
- The current sequence of `casanf16.shp` is: *A,B,**B**,C*,D,E,D,F,*A(S),B(S),**B(S)**,C(S)*,D(S),E(S),D(S),F(S)
  - The problem is so big that it's obvious, but for the sake of explanation, I still pasted a picture below.
  - And here, the DEF frames are using incorrect dimensions and positions, so they need to be handled together.

<img width="567" height="388" alt="image" src="https://github.com/user-attachments/assets/78f5dc3b-7906-4b80-b7b2-111fab7009ce" />

<img width="267" height="222" alt="image" src="https://github.com/user-attachments/assets/993fc93f-58ec-4261-958b-ae82cacadd95" />

*To ensure the robustness of the conclusion, it was reproduced through placement in the Map Editor in the current version CnCNetYR.*


> [!Tip]
> Regarding the necessity of handling the snow files as mentioned above: `xgxxxx.shp` is only used as a default image when other theaters do not have a corresponding image, so if you do not change those incorrect images, they will still use the original files instead of `xgxxxx.shp`, meaning your fix is still incomplete.

---
> I added files without "D" issue, however, for clsanf12,13,14 the problem was not confirmed in my tests

Pattern refers to the differences in assets under the *TEMPERATE* and *LUNAR* theaters compared to other climates. Specifically, the bridge surface uses the wrong texture. I see you have corrected the *TEMPERATE* version, so I hope you will correct the *LUNAR* version as well; otherwise, this fix is incomplete.

---
> I don't understand what should be done with the reduced-size images of buildings. In your opinion, they should be replaced? The Hollywood sign can also be considered a reduced size. Perhaps that was the intention.

Since you have isolated the scaled-down version of the Hollywood sign into a separate file, we are now able to resolve the inconsistencies in the original Hollywood sign. Otherwise, in the `[Map] -> Theater=SNOW` map, we would have *two small* Hollywood signs but *no large* one. The necessary migration at the map level to accommodate this change is described above.
- I have modified the `[CALA03]` image under *SNOW* theater and handled the adaptation of the map to this change, but due to the repository push restrictions set by your side, we are unable to directly add the changes to this PR.

CCCP84 commented:
> * but due to the repository push restrictions set by your side, we are unable to directly add the changes to this PR.

try now

CCCP84 committed:
North Korean Flag

CCCP84 commented:
> Pattern refers to the differences in assets under the _TEMPERATE_ and _LUNAR_ theaters compared to other climates. Specifically, the bridge surface uses the wrong texture. I see you have corrected the _TEMPERATE_ version, so I hope you will correct the _LUNAR_ version as well; otherwise, this fix is incomplete.

I already did it

CCCP84 committed:
casanf16

CCCP84 commented:
> The incorrect sequence of `casanf15.shp` is: A,B,_**B**_,C,A(S),B(S),_**B(S)**_,C(S)
> * The third frame of `[CASANF15]` is inconsistent with the versions of other theaters.

I don't understand what the frame sequence should be. I've checked several times, and the object is destroyed correctly in the game. How can I identify the problem?
In any case, please fix it as you see fit.

CCCP84 commented:
> I have not found any map with `[Map] -> Theater=SNOW` that contains `[CASANF16]`, so theoretically it can be safely replaced.

i replaced

CCCP84 commented:
> * The style of `gggateza.shp` in RA2:<img alt="image" width="176" height="127" src="https://private-user-images.githubusercontent.com/89088785/518701119-1fd9c5a5-a169-4111-be53-b030fc22ec24.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQxNDM0NjQsIm5iZiI6MTc2NDE0MzE2NCwicGF0aCI6Ii84OTA4ODc4NS81MTg3MDExMTktMWZkOWM1YTUtYTE2OS00MTExLWJlNTMtYjAzMGZjMjJlYzI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTI2VDA3NDYwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNiMzY2ZGFkMzllNGI3OWMwM2I1YzJiZTRjMGFiMjhmODhhY2ExZGIxYWU4ZjlkNTM0ZjM5YzU1NDE2ZjEyNDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ealX6CIlOfmGnHiZM5Gbjviz6hXX9DcZJffJdOv8MPQ">

Unfortunately, I don't understand what's written.
The screenshot shows g*gate_a
What do you suggest we do?

CCCP84 committed:
gaspot

CCCP84 commented:
the snow version of this asset, `gaspot.shp`, actually has a reduced image size (in fact, it retains the size from Tiberian Sun, so it is "reduced" compared to the non-snow version). Additional processing is required.

Fixed

CCCP84 commented:
@DeathFishAtEase You did a great job testing and identified many issues, but I didn't properly investigate and unreasonably rejected them at first. Sorry about that.
Thank you for participating.

DeathFishAtEase committed:
Correct the third frame of `casanf15.shp`

DeathFishAtEase committed:
Reconvert the `c?la03.shp` file using a PNG sequence.

Because rubble uses a different color palette, the other climates of `c?la03.shp` need to be converted separately and cannot directly use `cgla03.shp`.

CCCP84 reviewed (commented):
There was no need to add this for the Moon's desert and temperate because it all uses cgla03 in the absence. and it worked.
<img width="184" height="149" alt="image" src="https://github.com/user-attachments/assets/4e6dfb21-fd92-462f-a7b5-82197cf2a30f" />

only cala03 was needed

DeathFishAtEase commented:
> I already did it

I did not see you push the changes for `clsanf12.shp`, `clsanf13.shp`, and `clsanf14.shp`.

> I don't understand what the frame sequence should be. I've checked several times, and the object is destroyed correctly in the game. How can I identify the problem?
> In any case, please fix it as you see fit.

Changed

> Unfortunately, I don't understand what's written.
> The screenshot shows g*gate_a
> What do you suggest we do?

It may need discussion: for the east-west direction (screen ↖ to ↘ direction) gate, should we retain a TS legacy asset to make it usable for YR, or should we modify it to a new asset like the one Westwood has converted and applied to the north-south direction gate in RA2YR, which uses assets `g(a/g)gateza.shp` and `g(a/g)gate_b.shp` (this cannot be achieved through simple symmetry).

> the snow version of this asset, gaspot.shp, actually has a reduced image size (in fact, it retains the size from Tiberian Sun, so it is "reduced" compared to the non-snow version). Additional processing is required.
> 
> Fixed

I see you changed `gaspot.shp`, but you did not delete `ggspot.shp`; moreover, this BuildingType also has `gaspotmk.shp` and `gaspot_a.shp` which have the same issues as `gaspot.shp`, and you did not change them.

---

Finally, I want to confirm, have you noticed the rubble issue? I see that you have mostly changed civilian buildings (except for gate), so this is a very important issue.

> [!IMPORTANT]
> Their 4th frame is used as the rubble frame, and they use `iso~~~.pal` instead of `unit~~~.pal`. Unlike `unit~~~.pal`, `iso~~~.pal` has significant differences under different theaters.

In other words, `xgxxxx.shp` cannot be considered a general term.

Taking the current `c?tran03.shp` as an example, this is the current issue:

<img width="1820" height="1236" alt="image" src="https://github.com/user-attachments/assets/85d649d9-c470-4174-abb3-e182612dfb10" />

CCCP84 commented:
> I did not see you push the changes for `clsanf12.shp`, `clsanf13.shp`, and `clsanf14.shp`.

I already wrote above. I tested them. They are good.

CCCP84 commented:
> I see you changed `gaspot.shp`, but you did not delete `ggspot.shp`; moreover, this BuildingType also has `gaspotmk.shp` and `gaspot_a.shp` which have the same issues as `gaspot.shp`, and you did not change them.

Why delete ggspot.shp ? It used for all other theaters. gaspot.shp i added do solve size problem, you described.

CCCP84 commented:
> moreover, this BuildingType also has `gaspotmk.shp` and `gaspot_a.shp` which have the same issues as `gaspot.shp`, and you did not change them.

gaspot_a its just flashing lights or something like that, there can't be anything like that there.

CCCP84 commented:
> Their 4th frame is used as the rubble frame, and they use `iso~~~.pal` instead of `unit~~~.pal`. Unlike `unit~~~.pal`, `iso~~~.pal` has significant differences under different theaters.

so how can this be resolved if, for example, there was no option for a building in a particular theatre in principle?

CCCP84 commented:
In any case, I don't know how to fix these rubbles. So if you can do it, do it and lets finally merge PR

DeathFishAtEase commented:
> In any case, I don't know how to fix these rubbles. So if you can do it, do it and lets finally merge PR

Okay, I will continue to handle these issues.

> so how can this be resolved if, for example, there was no option for a building in a particular theatre in principle?

They need to be re-rendered to Shape files with the correct palette.

> Why delete ggspot.shp ? It used for all other theaters. gaspot.shp i added do solve size problem, you described.

I originally intended to directly replace it with the `unittem.pal` from TS to solve the related issues, because in ra2's `unittem.pal`, it simply replaced those two sets of colors representing the theater with a placeholder magenta. But later I thought, since we also have other palettes like `unitubn.pal`, I still chose to retain your original solution.

Furthermore, your existing approach of simply replacing `gaspot.shp` with `ggspot.shp` in your reply to this message is incorrect, because `gaspot.shp`, `ggspot.shp`, `gaspotmk.shp`, and `ggspotmk.shp` all use colors 208-239, which are called *theater colors*.

> [!NOTE]
> In TS, the two unit palettes use these color differences to apply mud or snow attachment effects to buildings.

In the `ggspot.shp` you submitted, the color indices used by the mud elements have been corrected to similar colors in ra2's `unittem.pal` that are not filled with placeholder magenta. However, in ra2's `unitsno.pal`, the theater colors are still the same as in TS, so it should only be scaled and not replaced with the content of `ggspot.shp`.

> gaspot_a its just flashing lights or something like that, there can't be anything like that there.

The *generic* versions of the three files are all **180x150**, while the *snow* versions of the three files are all **144x120** in size.

> I already wrote above. I tested them. They are good.

Perhaps it might be somewhat difficult to distinguish them with the naked eye, but I believe that by observing the following GIF that constantly switches between the correct version and the ***LUNAR*** version, you will understand the issue.

![Normal or LUNAR](https://github.com/user-attachments/assets/58e609e3-ab5b-4d16-afa4-feb6f9ad157a)
*`c?sanf14.shp` - the version used in `(all/sov)01umd.map` and the TEMPERATE/LUNAR version*

But it's okay, I will proceed to fix it.

DeathFishAtEase committed:
Fixed `clsanf12.shp`, `clsanf13.shp`, and `clsanf14.shp`

DeathFishAtEase committed:
Process the three files of `[GASPOT]` under SNOW theater and `ggspotmk.shp`.

DeathFishAtEase committed:
Correct the rubble palette for the small Hollywood sign.

DeathFishAtEase committed:
Correct the rubble palette for `[CASANF15]`

DeathFishAtEase committed:
Correct the rubble palette for `[CASANF16]`

DeathFishAtEase committed:
Correct the rubble palette for `[CASANF17]` & #15

#15: Re-render `casanf17.shp` to replace the version with incorrect dimensions.

DeathFishAtEase committed:
Correct the rubble palette for `[CASANF18]` & #16

#16: Re-render `casanf18.shp` to replace the version with incorrect dimensions.

DeathFishAtEase committed:
Correct the rubble palette for `caegyp01.shp`

In fact, this should be the graphic framework of `caegyp02.shp`, and as for the other three parts of the Great Pyramid, CAEGYP04-06 do not have SNOW versions, so CAEGYP01 does not have a SNOW version either. However, to avoid potential problems caused by this erroneous file, we still perform a simple processing on it.

DeathFishAtEase committed:
CALA14 & CALA15

DeathFishAtEase committed:
CALOND05

DeathFishAtEase committed:
CASANF05

DeathFishAtEase committed:
CASEAT02

DeathFishAtEase committed:
Correct the rubble palette for `ctlond06.shp`

In fact, this asset does not have versions for all theaters, but we’d better fill it in—when we start processing one of them, we will eventually have to complete it.

DeathFishAtEase committed:
CAMORR01-03

DeathFishAtEase committed:
Correct the rubble palette for `cdseat01.shp` & `clseat01.shp`

DeathFishAtEase committed:
Fix the size issue of `caseat01.shp`

It was originally larger than the standard size, same as the placeholder cubes in the -D and -L versions, to the extent that it did not conform to the `Foundation=2x2` range.

DeathFishAtEase committed:
CATRAN03

DeathFishAtEase committed:
[Second round check] Add missing `cuseat02.shp`

DeathFishAtEase commented:
> There is a small problem with the barrier: the barrier graphics is displayed over the graphics of the vehicles driving over it.

As mentioned above, it is an asset inherited from TS, which works fine under TS's `buildngz.sha`, but for RA2YR, even if you adjust its Z-depth via `NormalZAdjust` so that tanks are not blocked even in the northernmost cells, most of the image may be drawn below the ground due to the Z-depth being too far back. As for replacing `buildngz.sha`, it would be a global change to Z-depth and is incompatible with many existing buildings in RA2YR, so it is also not feasible.

So how does the RA2YR version of `[GAGATE_A]` avoid this? It lifts the pole, which shifts the lower boundary of the image's effective area upward and the right boundary leftward. If you use a non-zero index color to draw a dot at the lower-right corner of each frame, you will find that the westernmost pillar of `[GAGATE_A]`, which would originally be blocked by tanks driving above it, will instead block the tanks.

Obviously, GDI's `[GAGATE_B]` adopts an approach that lowers the 'gate' structure below the ground, which means that its lower and left boundaries in the image do not change, so it cannot use the same method as `[GAGATE_A]` in RA2YR to avoid this issue. Therefore, ***I suggest you abandon this Tiberian Sun legacy asset.***

Of course, this does not mean a death sentence for achieving this effect; it only discusses the incompatibility of this asset with RA2YR from a conventional approach. However, ***if you truly wish to use it in maps, you need to provide a compelling reason.***

After all, in my view: this is a TS legacy and is incomplete—the complete GDI gate requires paired use of buildings in both directions, and RA2YR has already repurposed one of them as a new gate. Thus, discarding it outright is the best option. Even if you choose not to discard it, any modifications should involve replacing the gate that still uses GDI elements with a new asset corresponding to the gate currently used in RA2YR, rather than continuing to rely on TS legacy content.

CCCP84 commented:
Any other opinions on this matter?

CCCP84 commented:
How can we restore a RA2 type gate?

DeathFishAtEase commented:
> Any other opinions on this matter?

I am checking whether other projects are completed, including but not limited to what others are expecting in the comments of this PR. To be honest, truly adapting all buildings to all theaters will be a huge project, so the subsequent processing is only limited to ensuring that the changed parts are successfully completed and the mentioned/promised/expected content is handled.

> How can we restore a RA2 type gate?

To achieve this, it would be necessary to remodel or hand-draw the ra2 gate in another direction.

- As I mentioned earlier, direct symmetry is not feasible because it's impossible to obtain the correct angle shadows through mirror images.
- Moreover, remodeling and rendering make it difficult to guarantee the same material textures. Even though I do have such assets here, there are still slight differences. In the Mod, it is certainly easy to replace all directional gates to maintain consistency, but doing so would also replace the original ra2yr gate with new assets, and for CnCNetYR, this is overstepping.
  - MO does have a symmetric version of the gate made by 4SG, but it belongs to the "exclusive to Mental Omega/not available for public download" category. And it actually also changes the original ra2yr gate.
- As for hand-drawing, it requires a lot of effort and is only suitable as a leisure activity. If someone has already done it before, then a project can directly borrow it, but during the project, specifically hand-drawing for one subtask is somewhat inefficient and not worthwhile.
  - I did have a half-finished product before, and on this issue, a friend of mine happened to express interest. Perhaps it will be completed later, but in the short term—at least it cannot be merged into this PR to complete it. If it can be completed, I might open a separate PR to add it, but I cannot guarantee the timing.

Therefore, from an engineering perspective, I recommend not to touch it.

---
Additionally, although I do not plan to improve the GDI gate in CnCNetYR, if you have your own YR Mod and want to add and use the complete two-way GDI gates, you can contact me to help you handle it. By the way, in TS, gates are actually military buildings that can be built by players, rather than being part of the Civil Structures category, which is the topic of this PR.

DeathFishAtEase committed:
Correct the rubble palette for `[YACOMD]`

Ensure that the yuri command center mentioned by RAZER in comment-3549797162 is available.

DeathFishAtEase commented:
> Nice job including the yuri command center. No telling how many times I've tried to use it in certain theaters but turned out invisible. Looking forward to this one 👍🏻

The original file changes in this PR did not include anything related to `[YACOMD]`. I just created 6 theaters' skirmish maps for testing and did not encounter this issue (both SteamYR & CnCNetYR). Since `ygcomd.shp` already exists in the original game`\ra2md.mix\genermd.mix\`, it shouldn't encounter this issue either, only the original rubble might have incorrect colors.
<img width="1600" height="1800" alt="YACOMD in 6 Theaters" src="https://github.com/user-attachments/assets/6dcd4172-d648-44fd-b42b-ef3ae57b1196" />
*Test whether `[YACOMD]` is displayed under 6 theaters in CnCNetYR using Custom Map.*

How did you encounter this issue? Was it in a scenario where the player manually built this Civil Building? I plan to address this issue together.

DeathFishAtEase committed:
Correct the rubble palette for `[CAEGYP02]`

Ensure that the pyramids mentioned by CCCP84 in comment-3555235201 is available.

DeathFishAtEase commented:
> How can we restore a RA2 type gate?

Good news! One of my friends became interested in this after knowing about it, so he used symmetrical body frames as new body frames, and supplemented shadow frames through modeling comparison in PS. I have obtained his authorization. What do you think of this effect?

![70328f201cc9a6713c51dc413df0ee29](https://github.com/user-attachments/assets/5c73cf88-4b42-49a8-8a42-f6b5c4da494f)

CCCP84 commented:
Great! Add this. We'll rename the gate from TS and merge it.

DeathFishAtEase committed:
Add the ra2 version of `[GAGATE_B]`

Co-authored-by: 王大可 <81621599+Duck-Wang@users.noreply.github.com>

DeathFishAtEase reviewed (approved):
Except for some issues mentioned in the discussion that cannot be reproduced and can be discussed later, this PR has met the expected standard.
